### PR TITLE
add full package path support when installing local add-ons

### LINKED ISSUE #746: allow full package path when installing add-ons
I was following the Chef server [upgrade documentation](https://docs.chef.io/upgrade_server.html). Once upgraded, I moved on and followed the steps to install the latest add-ons(manage, reporting, etc):

```
Then run the install command using the --path option to specify the location for the package:
$ chef-server-ctl install NAME_OF_PACKAGE --path /path/to/package
```

After downloading the required rpm's and reading the above note, I was confident this would:

```
chef-server-ctl install opscode-manage --path /media/chef/opscode-manage-1.11.0-1.el5.x86_64.rpm
```

The command failed with:
```
[2015-04-08T13:01:31+01:00] FATAL: Chef::Exceptions::Package: yum_package[opscode-manage] (private-chef::add_ons_local line 35) had an error: Chef::Exceptions::Package: Package opscode-manage not found:
```

It looks like the current implementation in the 'private-chef::add_ons_local' recipe is to expect the parent folder for the --path option.

I opened this issue to allow both the parent folder and the full package path to be specified for the --path option. This would be a more standard install method. I asked a few folks in order to validate this and they all felt for the full package path as I did.

### PR DESCRIPTION
**Pull request by alexpop**
fixes #746

### CODE REVIEW

File: files/private-chef-cookbooks/private-chef/recipes/add_ons_local.rb [hunk 1]
```diff
@@ -17,8 +17,19 @@
...
-  # find the newest package of each name
-  pkg_file = Dir["#{addon_path}/#{pkg}*.#{package_suffix}"].sort_by{ |f| File.mtime(f) }.last
+  if ::File.directory?(addon_path)
+    # find the newest package of each name if 'addon_path' is a directory
+    pkg_file = Dir["#{addon_path}/#{pkg}*.#{package_suffix}"].sort_by{ |f| File.mtime(f) }.last
+  else
+    if File.basename(addon_path) =~ /#{pkg}/
+      # use the full package path if it contains the package name
+      pkg_file = addon_path
+    else
```
Comment (stevendanna) [inline files/private-chef-cookbooks/private-chef/recipes/add_ons_local.rb]: Why not just assume we have the right file if someone passes in a full path, perhaps with a ::File.exists? test?  Looking at the parent directory doesn't seem like intuitive behavior to me.

Comment (alexpop) [inline files/private-chef-cookbooks/private-chef/recipes/add_ons_local.rb]: Thanks for the review @stevendanna.
The reason I'm verifying that 'pkg' is part of the file name is that node['private_chef']['addons']['packages'] comes as an array. Hence the original "node['private_chef']['addons']['packages'].each" loop.
Here's the output:

```
[root@ap-local-chef-server-12 chef]# chef-server-ctl install opscode-reporting --path /media/sf_mnt_kits/chef/opscode-reporting-1.3.0-1.el6.x86_64.rpm
---SNIP---
Recipe: private-chef::add_ons_local
  * yum_package[opscode-manage] action install (up to date)
  * yum_package[opscode-reporting] action install (up to date)
---SNIP---
Chef Client finished, 0/2 resources updated in 2.802826526 seconds
```

Without the parent directory check, it tries to install 'opscode-manage' using the full path of the reporting rpm.
I don't know why we have an array of packages. Is it an attribute cache/merging issue? I've also noticed  "node['private_chef']['addons']['path']" getting stuck to '/media/sf_mnt_kits/chef/' even if I run the install without --path, like this:

```
[root@ap-local-chef-server-12 chef]# chef-server-ctl install chef-sync
---SNIP---

From: /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/recipes/add_ons_local.rb @ line 19 Chef::Mixin::FromFile#from_file:

    18: require "pry"
 => 19: binding.pry
    20:
    21: addon_path = node['private_chef']['addons']['path']
    22:
    23: node['private_chef']['addons']['packages'].each do |pkg|

[1] pry(#<Chef::Recipe>)> node['private_chef']['addons']['path']
=> "/media/sf_mnt_kits/chef/opscode-reporting-1.3.0-1.el6.x86_64.rpm"
[2] pry(#<Chef::Recipe>)> node['private_chef']['addons']['packages']
=> ["opscode-manage", "opscode-reporting", "chef-sync"]
[3] pry(#<Chef::Recipe>)> quit
Converging 6 resources
Recipe: private-chef::add_ons_wrapper
  * ruby_block[addon_install_notification_opscode-manage] action nothing (skipped due to action :nothing)
  * ruby_block[addon_install_notification_opscode-reporting] action nothing (skipped due to action :nothing)
  * ruby_block[addon_install_notification_chef-sync] action nothing (skipped due to action :nothing)
Recipe: private-chef::add_ons_local
  * yum_package[opscode-manage] action install (up to date)
  * yum_package[opscode-reporting] action install (up to date)
  * yum_package[chef-sync] action install
    * Package chef-sync not found:
---SNIP---
[2015-04-08T14:33:21+01:00] FATAL: Chef::Exceptions::Package: yum_package[chef-sync] (private-chef::add_ons_local line 38) had an error: Chef::Exceptions::Package: Package chef-sync not found:
```

I've looked into the attributes file used by the chef-client run:
```
/opt/opscode/embedded/bin/ruby /opt/opscode/embedded/bin/chef-client -z --config /opt/opscode/embedded/cookbooks/solo.rb --json-attributes /opt/opscode/embedded/cookbooks/install_params.json --log_level fatal
```
and the path should be null:
```
[root@ap-local-chef-server-12 ~]# cat /opt/opscode/embedded/cookbooks/install_params.json
{"run_list":["recipe[private-chef::add_ons_wrapper]"],"private_chef":{"addons":{"install":true,"packages":["chef-sync"],"path":null}}}
```

**Comment by stevendanna** [comment]
@alexpop Are you on 12.0.6?  I think part of this is being influenced by https://github.com/chef/omnibus-ctl/issues/24 which should be fixed in the latest release.

**Comment by alexpop** [comment]
@stevendanna, my version is 12.0.5:
```
[root@ap-local-chef-server-12 chef]# head /opt/opscode/version-manifest.txt
chef-server 12.0.5
```

I'll upgrade to the latest and update this thread. Thanks

**Comment by alexpop** [comment]
@stevendanna, @sean-horn, same issue on chef-server 12.0.7 

```
[root@ap-local-chef-server-12 chef]# head -1 /opt/opscode/version-manifest.txt 
chef-server 12.0.7

[root@ap-local-chef-server-12 chef]# chef-server-ctl install opscode-manage
---SNIP---
[1] pry(#<Chef::Recipe>)> node['private_chef']['addons']['packages']
=> ["opscode-reporting", "opscode-manage"]
[2] pry(#<Chef::Recipe>)> node['private_chef']['addons']['path']
=> "/media/sf_mnt_kits/chef"
```

It looks like the state is stored in:
/opt/opscode/embedded/nodes/ap-local-chef-server-12.opschef.tv.json

**Comment by stevendanna** [comment]
Ufffffff. We nuke the node state before the reconfigure command, but I forgot about the install command:  https://github.com/chef/omnibus-ctl/commit/0fabcd382fa0926e1d8a07bcfa91e4ed4e3fdad0

**Comment by stevendanna** [comment]
Grumble grumble grumble grumble.  Let's get a separate chef-server bug opened for this (the node state issue during install).

**Comment by alexpop** [comment]
Ok, I opened: https://github.com/chef/omnibus-ctl/issues/29
Once this is fixed, I'll update my changes for issue #748

**Comment by stevendanna** [comment]
This was moved to https://github.com/chef/chef-server/pull/173

**Pull request closed without merging**